src/permutations.jl

```diff
@@ -9,14 +9,52 @@ export
     permutations
 
 
-struct Permutations{T}
-    a::T
-    t::Int
+struct PermutationIterator{T}
+    data::T
+    length::Int
+end
+
+function has_repeats(state)
+    for outer in 1:length(state)
+        for inner in (outer+1):length(state)
+            if state[outer] == state[inner]
+                return true
+            end
+        end
+    end
+    return false
+end
+
+function increment!(state, max)
+    state[end] += 1
+    for i in length(state):-1:2
+        if state[i] > max
+            state[i] = 1
+            state[i-1] += 1
+        end
+    end
+end
+
+function next_permutation!(state, max)
+    while true
+        increment!(state, max)
+        if !has_repeats(state)
+            break
+        end
+    end
+end
+
+function Base.iterate(p::PermutationIterator, state=ones(Int, p.length))
```

Comment:
code should add an assertion that `state` uses one based indexing (or change the code not to assume one based indexing). also in all code, where possible I would add type annotations to make sure no unwanted values are passed.

Reply:
Looks like my fix of using `Base.require_one_based_indexing` won't work with Julia 1.0. 

Is there another way to check that data structures use 1-based indexing?

Reply:
use `firstindex`, it is supported on Julia 1.0

Reply:
As of commit [5deef9c](https://github.com/JuliaMath/Combinatorics.jl/pull/122/commits/5deef9ccebe86013aaa13b4c7f266bcea03a681a), have attempted to make it work for OffsetArrays too. However, running into issues that I don't yet understand. Of the two tests I wrote for OffsetArrays, one is broken, and the other hits an infinite loop.

In the original code, I interpreted `length(state)` to mean "get me the last index". However, I'm not sure if that is always correct. @rdeits, do you remember if that's what you meant when you originally wrote this code?

Reply:
I have corrected the code, but I did it in my head so there might be mistakes.

Reply:
Your suggestions made all the tests pass